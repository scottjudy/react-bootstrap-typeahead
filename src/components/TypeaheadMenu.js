// @flow

import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import type { Node } from 'react';

import Highlighter from './Highlighter';
import Menu from './Menu';
import MenuItem from './MenuItem';

import { getOptionLabel, getOptionProperty } from '../utils';

import type { MenuComponentProps } from './Menu';
import type { LabelKey, Option } from '../types';

export type TypeaheadMenuProps = MenuComponentProps & {
  labelKey: LabelKey,
  newSelectionPrefix: Node,
  options: Option[],
  paginationText: Node,
  renderMenuItemChildren: (Option, TypeaheadMenuProps, number) => Node,
};

const propTypes = {
  /**
   * Provides the ability to specify a prefix before the user-entered text to
   * indicate that the selection will be new. No-op unless `allowNew={true}`.
   */
  newSelectionPrefix: PropTypes.node,
  /**
   * Prompt displayed when large data sets are paginated.
   */
  paginationText: PropTypes.node,
  /**
   * Provides a hook for customized rendering of menu item contents.
   */
  renderMenuItemChildren: PropTypes.func,
};

const defaultProps = {
  newSelectionPrefix: 'New selection: ',
  paginationText: 'Display additional results...',
  renderMenuItemChildren: (
    option: Option,
    props: TypeaheadMenuProps,
    idx: number
  ) => (
    <Highlighter search={props.text}>
      {getOptionLabel(option, props.labelKey)}
    </Highlighter>
  ),
};

const TypeaheadMenu = (props: TypeaheadMenuProps) => {
  const {
    labelKey,
    newSelectionPrefix,
    options,
    paginationText,
    renderMenuItemChildren,
    text,
    ...menuProps
  } = props;

  const renderMenuItem = (option: Option, position: number) => {
    const label = getOptionLabel(option, labelKey);

    const menuItemProps = {
      disabled: getOptionProperty(option, 'disabled'),
      label,
      option,
      position,
    };

    if (option.customOption) {
      return (
        <MenuItem
          {...menuItemProps}
          className="rbt-menu-custom-option"
          key={position}
          label={label}>
          {newSelectionPrefix}
          <Highlighter search={text}>{label}</Highlighter>
        </MenuItem>
      );
    }

    if (option.paginationOption) {
      return (
        <Fragment key="pagination-item">
          <Menu.Divider />
          <MenuItem
            {...menuItemProps}
            className="rbt-menu-pagination-option"
            label={paginationText}>
            {paginationText}
          </MenuItem>
        </Fragment>
      );
    }

    return (
      <MenuItem {...menuItemProps} key={position}>
        {renderMenuItemChildren(option, props, position)}
      </MenuItem>
    );
  };

  return (
    // Explictly pass `text` so Flow doesn't complain...
    <Menu {...menuProps} text={text}>
      {options.map(renderMenuItem)}
    </Menu>
  );
};

TypeaheadMenu.propTypes = propTypes;
TypeaheadMenu.defaultProps = defaultProps;

export default TypeaheadMenu;
