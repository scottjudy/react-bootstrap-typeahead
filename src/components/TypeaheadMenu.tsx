import PropTypes from 'prop-types';
import React, { ReactNode } from 'react';

import Highlighter from './Highlighter';
import Menu, { MenuProps } from './Menu';
import MenuItem from './MenuItem';

import { getOptionLabel, getOptionProperty, isString } from '../utils';
import { LabelKey, Option } from '../types';

export type RenderMenuItemChildren = (
  option: Option,
  menuProps: TypeaheadMenuProps,
  idx: number
) => JSX.Element;

export interface TypeaheadMenuProps extends Omit<MenuProps, 'ref'> {
  labelKey: LabelKey;
  newSelectionPrefix: ReactNode;
  options: Option[];
  paginationText: ReactNode;
  renderMenuItemChildren: RenderMenuItemChildren;
  text: string;
}

const propTypes = {
  /**
   * Provides the ability to specify a prefix before the user-entered text to
   * indicate that the selection will be new. No-op unless `allowNew={true}`.
   */
  newSelectionPrefix: PropTypes.node,
  /**
   * Prompt displayed when large data sets are paginated.
   */
  paginationText: PropTypes.node,
  /**
   * Provides a hook for customized rendering of menu item contents.
   */
  renderMenuItemChildren: PropTypes.func,
};

const defaultProps = {
  newSelectionPrefix: 'New selection: ',
  paginationText: 'Display additional results...',
  renderMenuItemChildren: (option: Option, props: TypeaheadMenuProps) => (
    <Highlighter search={props.text}>
      {getOptionLabel(option, props.labelKey)}
    </Highlighter>
  ),
};

const TypeaheadMenu = (props: TypeaheadMenuProps) => {
  const {
    labelKey,
    newSelectionPrefix,
    options,
    paginationText,
    renderMenuItemChildren,
    ...menuProps
  } = props;

  const renderMenuItem = (option: Option, position: number) => {
    const label = getOptionLabel(option, labelKey);

    const menuItemProps = {
      disabled: !!getOptionProperty(option, 'disabled'),
      label,
      option,
      position,
    };

    if (getOptionProperty(option, 'customOption')) {
      return (
        <MenuItem
          {...menuItemProps}
          className="rbt-menu-custom-option"
          key={position}
          label={label}>
          {newSelectionPrefix}
          <Highlighter search={menuProps.text}>{label}</Highlighter>
        </MenuItem>
      );
    }

    if (getOptionProperty(option, 'paginationOption')) {
      return (
        <>
          <Menu.Divider />
          <MenuItem
            {...menuItemProps}
            className="rbt-menu-pagination-option"
            label={
              // TODO: Fix how (aria-)labels are passed to `MenuItem`.
              // `paginationText` can be a ReactNode.
              isString(paginationText) ? paginationText : ''
            }>
            {paginationText}
          </MenuItem>
        </>
      );
    }

    return (
      <MenuItem {...menuItemProps} key={position}>
        {renderMenuItemChildren(option, props, position)}
      </MenuItem>
    );
  };

  return <Menu {...menuProps}>{options.map(renderMenuItem)}</Menu>;
};

TypeaheadMenu.propTypes = propTypes;
TypeaheadMenu.defaultProps = defaultProps;

export default TypeaheadMenu;
