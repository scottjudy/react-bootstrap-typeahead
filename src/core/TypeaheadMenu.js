import {pick} from 'lodash';
import React from 'react';

import Overlay from './Overlay';

const TypeaheadMenu = (props) => {
  const {children, isMenuShown, menuId, results} = props;

  const overlayProps = pick(props, [
    'align',
    'className',
    'container',
    'dropup',
    'flip',
    'onMenuHide',
    'onMenuShow',
    'onMenuToggle',
    'referenceElement',
  ]);

  const menuProps = pick(props, [
    'emptyLabel',
    'labelKey',
    'maxHeight',
    'newSelectionPrefix',
    'renderMenuItemChildren',
    'text',
  ]);

  menuProps.id = menuId;

  return (
    <Overlay
      {...overlayProps}
      show={isMenuShown}>
      {children(results, menuProps)}
    </Overlay>
  );
};

export default TypeaheadMenu;
