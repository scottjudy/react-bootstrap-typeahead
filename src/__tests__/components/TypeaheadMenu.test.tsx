import React from 'react';

import TypeaheadMenu from '../../components/TypeaheadMenu';

import options from '../data';
import {
  getItems,
  getMenu,
  noop,
  prepareSnapshot,
  render,
  screen,
  TestProvider,
} from '../helpers';

const TestComponent = (props) => (
  <TestProvider selected={[]}>
    {({ state }) => (
      <TypeaheadMenu
        {...state}
        id="menu-id"
        labelKey="name"
        options={options}
        text=""
        {...props}
      />
    )}
  </TestProvider>
);

describe('<TypeaheadMenu>', () => {
  it('renders a snapshot', () => {
    expect(prepareSnapshot(<TestComponent />)).toMatchSnapshot();
  });

  it('renders a menu with the specified max-height', () => {
    const { rerender } = render(<TestComponent maxHeight="200px" />);
    expect(getMenu().getAttribute('style')).toContain('max-height: 200px;');

    rerender(<TestComponent maxHeight="50%" />);
    expect(getMenu().getAttribute('style')).toContain('max-height: 50%;');
  });

  it('renders an empty state when there are no results', () => {
    const emptyLabel = 'No matches found.';

    render(<TestComponent emptyLabel={emptyLabel} options={[]} />);
    const menuItems = getItems();

    expect(menuItems).toHaveLength(1);
    expect(menuItems[0]).toHaveTextContent(emptyLabel);
  });

  describe('pagination behaviors', () => {
    let paginationProps, paginationText;

    beforeEach(() => {
      paginationText = 'More results...';
      paginationProps = {
        options: options.concat({
          name: '',
          paginationOption: true,
        }),
        paginationText,
      };
    });

    it('displays a paginator', () => {
      render(<TestComponent {...paginationProps} />);
      expect(screen.queryByText(paginationText)).toBeInTheDocument();
    });

    it('does not show a paginator when there are no results', () => {
      render(<TestComponent {...paginationProps} options={[]} />);
      expect(screen.queryByText(paginationText)).not.toBeInTheDocument();
    });
  });
});
