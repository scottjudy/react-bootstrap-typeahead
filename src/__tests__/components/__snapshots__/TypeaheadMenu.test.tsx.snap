// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TypeaheadMenu> renders a snapshot 1`] = `
<div
  aria-label="menu-options"
  className="rbt-menu dropdown-menu show"
  id="menu-id"
  onMouseDown={[Function]}
  role="listbox"
  style={
    Object {
      "display": "block",
      "maxHeight": "300px",
      "overflow": "auto",
    }
  }
>
  <a
    aria-label="Alabama"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-0"
    label="Alabama"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Montgomery",
        "name": "Alabama",
        "population": 4780127,
        "region": "South",
      }
    }
    position={0}
    role="option"
  >
    Alabama
  </a>
  <a
    aria-label="Alaska"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-1"
    label="Alaska"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Juneau",
        "name": "Alaska",
        "population": 710249,
        "region": "West",
      }
    }
    position={1}
    role="option"
  >
    Alaska
  </a>
  <a
    aria-label="Arizona"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-2"
    label="Arizona"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Phoenix",
        "name": "Arizona",
        "population": 6392307,
        "region": "West",
      }
    }
    position={2}
    role="option"
  >
    Arizona
  </a>
  <a
    aria-label="Arkansas"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-3"
    label="Arkansas"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Little Rock",
        "name": "Arkansas",
        "population": 2915958,
        "region": "South",
      }
    }
    position={3}
    role="option"
  >
    Arkansas
  </a>
  <a
    aria-label="California"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-4"
    label="California"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Sacramento",
        "name": "California",
        "population": 37254503,
        "region": "West",
      }
    }
    position={4}
    role="option"
  >
    California
  </a>
  <a
    aria-label="Colorado"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-5"
    label="Colorado"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Denver",
        "name": "Colorado",
        "population": 5029324,
        "region": "West",
      }
    }
    position={5}
    role="option"
  >
    Colorado
  </a>
  <a
    aria-label="Connecticut"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-6"
    label="Connecticut"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Hartford",
        "name": "Connecticut",
        "population": 3574118,
        "region": "Northeast",
      }
    }
    position={6}
    role="option"
  >
    Connecticut
  </a>
  <a
    aria-label="Delaware"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-7"
    label="Delaware"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Dover",
        "name": "Delaware",
        "population": 897936,
        "region": "South",
      }
    }
    position={7}
    role="option"
  >
    Delaware
  </a>
  <a
    aria-label="Florida"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-8"
    label="Florida"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Tallahassee",
        "name": "Florida",
        "population": 18804623,
        "region": "South",
      }
    }
    position={8}
    role="option"
  >
    Florida
  </a>
  <a
    aria-label="Georgia"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-9"
    label="Georgia"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Atlanta",
        "name": "Georgia",
        "population": 9688681,
        "region": "South",
      }
    }
    position={9}
    role="option"
  >
    Georgia
  </a>
  <a
    aria-label="Hawaii"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-10"
    label="Hawaii"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Honolulu",
        "name": "Hawaii",
        "population": 1360301,
        "region": "West",
      }
    }
    position={10}
    role="option"
  >
    Hawaii
  </a>
  <a
    aria-label="Idaho"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-11"
    label="Idaho"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Boise",
        "name": "Idaho",
        "population": 1567652,
        "region": "West",
      }
    }
    position={11}
    role="option"
  >
    Idaho
  </a>
  <a
    aria-label="Illinois"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-12"
    label="Illinois"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Springfield",
        "name": "Illinois",
        "population": 12831549,
        "region": "Midwest",
      }
    }
    position={12}
    role="option"
  >
    Illinois
  </a>
  <a
    aria-label="Indiana"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-13"
    label="Indiana"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Indianapolis",
        "name": "Indiana",
        "population": 6484229,
        "region": "Midwest",
      }
    }
    position={13}
    role="option"
  >
    Indiana
  </a>
  <a
    aria-label="Iowa"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-14"
    label="Iowa"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Des Moines",
        "name": "Iowa",
        "population": 3046869,
        "region": "Midwest",
      }
    }
    position={14}
    role="option"
  >
    Iowa
  </a>
  <a
    aria-label="Kansas"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-15"
    label="Kansas"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Topeka",
        "name": "Kansas",
        "population": 2853132,
        "region": "Midwest",
      }
    }
    position={15}
    role="option"
  >
    Kansas
  </a>
  <a
    aria-label="Kentucky"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-16"
    label="Kentucky"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Frankfort",
        "name": "Kentucky",
        "population": 4339349,
        "region": "South",
      }
    }
    position={16}
    role="option"
  >
    Kentucky
  </a>
  <a
    aria-label="Louisiana"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-17"
    label="Louisiana"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Baton Rouge",
        "name": "Louisiana",
        "population": 4533479,
        "region": "South",
      }
    }
    position={17}
    role="option"
  >
    Louisiana
  </a>
  <a
    aria-label="Maine"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-18"
    label="Maine"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Augusta",
        "name": "Maine",
        "population": 1328361,
        "region": "Northeast",
      }
    }
    position={18}
    role="option"
  >
    Maine
  </a>
  <a
    aria-label="Maryland"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-19"
    label="Maryland"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Annapolis",
        "name": "Maryland",
        "population": 5773785,
        "region": "South",
      }
    }
    position={19}
    role="option"
  >
    Maryland
  </a>
  <a
    aria-label="Massachusetts"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-20"
    label="Massachusetts"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Boston",
        "name": "Massachusetts",
        "population": 6547817,
        "region": "Northeast",
      }
    }
    position={20}
    role="option"
  >
    Massachusetts
  </a>
  <a
    aria-label="Michigan"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-21"
    label="Michigan"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Lansing",
        "name": "Michigan",
        "population": 9884129,
        "region": "Midwest",
      }
    }
    position={21}
    role="option"
  >
    Michigan
  </a>
  <a
    aria-label="Minnesota"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-22"
    label="Minnesota"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Saint Paul",
        "name": "Minnesota",
        "population": 5303925,
        "region": "Midwest",
      }
    }
    position={22}
    role="option"
  >
    Minnesota
  </a>
  <a
    aria-label="Mississippi"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-23"
    label="Mississippi"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Jackson",
        "name": "Mississippi",
        "population": 2968103,
        "region": "South",
      }
    }
    position={23}
    role="option"
  >
    Mississippi
  </a>
  <a
    aria-label="Missouri"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-24"
    label="Missouri"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Jefferson City",
        "name": "Missouri",
        "population": 5988927,
        "region": "Midwest",
      }
    }
    position={24}
    role="option"
  >
    Missouri
  </a>
  <a
    aria-label="Montana"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-25"
    label="Montana"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Alberta",
        "name": "Montana",
        "population": 989417,
        "region": "West",
      }
    }
    position={25}
    role="option"
  >
    Montana
  </a>
  <a
    aria-label="Nebraska"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-26"
    label="Nebraska"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Lincoln",
        "name": "Nebraska",
        "population": 1826341,
        "region": "Midwest",
      }
    }
    position={26}
    role="option"
  >
    Nebraska
  </a>
  <a
    aria-label="Nevada"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-27"
    label="Nevada"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Carson City",
        "name": "Nevada",
        "population": 2700691,
        "region": "West",
      }
    }
    position={27}
    role="option"
  >
    Nevada
  </a>
  <a
    aria-label="New Hampshire"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-28"
    label="New Hampshire"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Concord",
        "name": "New Hampshire",
        "population": 1316466,
        "region": "Northeast",
      }
    }
    position={28}
    role="option"
  >
    New Hampshire
  </a>
  <a
    aria-label="New Jersey"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-29"
    label="New Jersey"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Trenton",
        "name": "New Jersey",
        "population": 8791936,
        "region": "Northeast",
      }
    }
    position={29}
    role="option"
  >
    New Jersey
  </a>
  <a
    aria-label="New Mexico"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-30"
    label="New Mexico"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Santa Fe",
        "name": "New Mexico",
        "population": 2059192,
        "region": "West",
      }
    }
    position={30}
    role="option"
  >
    New Mexico
  </a>
  <a
    aria-label="New York"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-31"
    label="New York"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Albany",
        "name": "New York",
        "population": 19378087,
        "region": "Northeast",
      }
    }
    position={31}
    role="option"
  >
    New York
  </a>
  <a
    aria-label="North Carolina"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-32"
    label="North Carolina"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Raleigh",
        "name": "North Carolina",
        "population": 9535692,
        "region": "South",
      }
    }
    position={32}
    role="option"
  >
    North Carolina
  </a>
  <a
    aria-label="North Dakota"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-33"
    label="North Dakota"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Bismarck",
        "name": "North Dakota",
        "population": 672591,
        "region": "Midwest",
      }
    }
    position={33}
    role="option"
  >
    North Dakota
  </a>
  <a
    aria-label="Ohio"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-34"
    label="Ohio"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Columbus",
        "name": "Ohio",
        "population": 11536725,
        "region": "Midwest",
      }
    }
    position={34}
    role="option"
  >
    Ohio
  </a>
  <a
    aria-label="Oklahoma"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-35"
    label="Oklahoma"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Oklahoma City",
        "name": "Oklahoma",
        "population": 3751616,
        "region": "South",
      }
    }
    position={35}
    role="option"
  >
    Oklahoma
  </a>
  <a
    aria-label="Oregon"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-36"
    label="Oregon"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Salem",
        "name": "Oregon",
        "population": 3831073,
        "region": "West",
      }
    }
    position={36}
    role="option"
  >
    Oregon
  </a>
  <a
    aria-label="Pennsylvania"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-37"
    label="Pennsylvania"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Harrisburg",
        "name": "Pennsylvania",
        "population": 12702887,
        "region": "Northeast",
      }
    }
    position={37}
    role="option"
  >
    Pennsylvania
  </a>
  <a
    aria-label="Rhode Island"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-38"
    label="Rhode Island"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Providence",
        "name": "Rhode Island",
        "population": 1052931,
        "region": "Northeast",
      }
    }
    position={38}
    role="option"
  >
    Rhode Island
  </a>
  <a
    aria-label="South Carolina"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-39"
    label="South Carolina"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Columbia",
        "name": "South Carolina",
        "population": 4625401,
        "region": "South",
      }
    }
    position={39}
    role="option"
  >
    South Carolina
  </a>
  <a
    aria-label="South Dakota"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-40"
    label="South Dakota"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Pierre",
        "name": "South Dakota",
        "population": 814191,
        "region": "Midwest",
      }
    }
    position={40}
    role="option"
  >
    South Dakota
  </a>
  <a
    aria-label="Tennessee"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-41"
    label="Tennessee"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Nashville",
        "name": "Tennessee",
        "population": 6346275,
        "region": "South",
      }
    }
    position={41}
    role="option"
  >
    Tennessee
  </a>
  <a
    aria-label="Texas"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-42"
    label="Texas"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Austin",
        "name": "Texas",
        "population": 25146105,
        "region": "South",
      }
    }
    position={42}
    role="option"
  >
    Texas
  </a>
  <a
    aria-label="Utah"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-43"
    label="Utah"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Salt Lake City",
        "name": "Utah",
        "population": 2763888,
        "region": "West",
      }
    }
    position={43}
    role="option"
  >
    Utah
  </a>
  <a
    aria-label="Vermont"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-44"
    label="Vermont"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Montpelier",
        "name": "Vermont",
        "population": 625745,
        "region": "Northeast",
      }
    }
    position={44}
    role="option"
  >
    Vermont
  </a>
  <a
    aria-label="Virginia"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-45"
    label="Virginia"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Richmond",
        "name": "Virginia",
        "population": 8001045,
        "region": "South",
      }
    }
    position={45}
    role="option"
  >
    Virginia
  </a>
  <a
    aria-label="Washington"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-46"
    label="Washington"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Olympia",
        "name": "Washington",
        "population": 6724543,
        "region": "West",
      }
    }
    position={46}
    role="option"
  >
    Washington
  </a>
  <a
    aria-label="West Virginia"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-47"
    label="West Virginia"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Charleston",
        "name": "West Virginia",
        "population": 1853011,
        "region": "South",
      }
    }
    position={47}
    role="option"
  >
    West Virginia
  </a>
  <a
    aria-label="Wisconsin"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-48"
    label="Wisconsin"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Madison",
        "name": "Wisconsin",
        "population": 5687289,
        "region": "West",
      }
    }
    position={48}
    role="option"
  >
    Wisconsin
  </a>
  <a
    aria-label="Wyoming"
    aria-selected={false}
    className="dropdown-item"
    href="#"
    id="-item-49"
    label="Wyoming"
    onClick={[Function]}
    onMouseDown={[Function]}
    option={
      Object {
        "capital": "Cheyenne",
        "name": "Wyoming",
        "population": 563767,
        "region": "West",
      }
    }
    position={49}
    role="option"
  >
    Wyoming
  </a>
</div>
`;
